Support for logging via rcutils

dhood opened the pull request:
connects to https://github.com/ros2/rclpy/issues/103

Could I get some feedback on the direction this is going in? I add `logging` as an additional module of the `rclpy` package but it might belong in its own package or in the `rclpy` module itself.

wjwwood commented:
Looks reasonable to me so far. It could either be its own module or part of the existing one. I don't think it needs to be in a separate package. Is there any downside to initializing it automatically?

dirk-thomas commented on rclpy/CMakeLists.txt:
```diff
@@ -74,6 +74,38 @@ install(TARGETS ${PROJECT_NAME}
+target_include_directories(${library_name}
+  PUBLIC
+  ${PythonExtra_INCLUDE_DIRS}
+)
```
Instead of duplicating the logic to update the target with various information this should be moved into a function for reuse.

dirk-thomas commented on rclpy/src/rclpy/_rclpy_logging.c:
```diff
@@ -0,0 +1,90 @@
+static struct PyModuleDef _rclpy_logging_module = {
+  PyModuleDef_HEAD_INIT,
+  "_rclpy_logging",
+  "_rclpy_logging_doc",
```
Instead of a dummy this should be `NULL` or a actual docstring.

dirk-thomas commented on rclpy/test/test_logging.py:
```diff
@@ -0,0 +1,36 @@
+    def setUpClass(cls):
+        rclpy.logging.initialize()
+
+    def test_set_get_severity_threshold(self):
```
Nitpick about the name: Use `*_get_set_*` (since that is the order used in the test) or just `test_severity_threshold`.

dhood commented:
Putting this into review.

Summary:
Log calls can be on a specific user-created logger, or the general `rclpy.logging.log()` calls go to the root rclpy logger.

At some point the C extension was wrapping calls to each supported logging macro from rcutils to avoid reimplementing each logging "feature" (e.g. throttle, skip_first). However, this would lead to all rclpy logging calls sharing the same 'context' (e.g. last throttle time), so the C extension now just wraps the underlying `rcutils_log` function call and rclpy doesn't reuse any of the implementation of logging "features" from rcutils except 'named'.

Each log call has a context associated with it that is initialised on the first call. Relevant logging features (referred to as "filters" in rclpy to match python's logging module) for the call are detected from the combination of keyword arguments in the log call, and each filter has an opportunity to add their required state variables to the context. On each log call the filters check the context to see if logging should be skipped or not.

To match behaviour in the C implementation, no changes are permitted to any logging options on-the-fly for a particular log call. There's some overhead involved in performing this check.

Regarding initialisation, each log call uses `RCUTILS_LOGGING_AUTOINIT`, and as of https://github.com/ros2/rcutils/pull/52 there's no need to manually initialise before setting the severity. The initialisation doesn't happen at module import time anymore because @dirk-thomas pointed out that someone might be interested in importing the module, customising things, then calling initialise themselves.

The usage is captured in the unit tests. Note that I added a `return_log_condition` option solely for the purpose of simplifying testing. It causes the log call to return whether or not the message got logged, or if one of the filters filtered it out. That might not be best practice, it was just simpler than checking everything with launch_testing: I'm open to other alternatives.

Todos:
- Currently all loggers are using the global severity threshold from rcutils, but I think that should be changed to an attribute of the logger. At the moment it'd be independent of the rcutils logging library.
- All of the convenience functions take the same set of filter options but none of them have been documented which inhibits autocompletion.
- 'function' and/or 'expression' filter support

wjwwood commented:
> Currently all loggers are using the global severity threshold from rcutils, but I think that should be changed to an attribute of the logger. At the moment it'd be independent of the rcutils logging library.

That's a feature that I have pointed out before should be supported by the common logging API (so in `rcutils`. I believe it wasn't implemented yet because you need some sort of map like structure to correlate settings to a given logger name and we didn't have that in C. I proposed the implementation could use C++ and the `std::map` structure for the storage and lookup of the settings by logger name, but I think @dirk-thomas wanted to use C and also did not want to implement the data structure. So we left it hanging waiting on a decision about which C library to pull it that would give us that data structure.

You could implement it in python separately, but I think that's wasted effort if we know it needs to work at the rcutils level as well in the future.

> Allow log calls to take a format string + args (currently only pre-formatted strings are supported)

I think that's fine for Python since all of its string formatting options happen in place.

dhood commented:
note to reviewers: something in the logic for [determining the caller's filename etc](https://github.com/ros2/rclpy/pull/102/files#diff-4c2d5212f21dae17a48ecd1a1dce3c24R30) isn't working correctly on windows and I am looking into it

dirk-thomas commented:
I am not sure if the `return_log_condition` option is necessary. It would be fine of the `log` function simply always returns a boolean indicating of the C log function was called. The user can just commonly ignore the return value.

dhood commented:
> Allow log calls to take a format string + args (currently only pre-formatted strings are supported)
I think that's fine for Python since all of its string formatting options happen in place.

@wjwwood Do you mean that we don't want to add support for `log(format_string, *args, **kwargs)` like in ROS 1 and python's logging module? (in my mind it's a good thing to add for compatibility, but is not immediately necessary)

dhood commented:
>  It would be fine of the log function simply always returns a boolean indicating of the C log function was called. The user can just commonly ignore the return value.

@dirk-thomas I forgot to mention that the reason why I added it was because if it always returns the value, then logging in an interactive environment shows the return value which seemed weird, e.g.:

```
In [3]: import rclpy.logging

In [4]: rclpy.logging.log('asdf', 1, once=True, throttle_duration_sec=1)
[INFO] []: asdf (<module>() at /home/dhood/src/ros2ci/<ipython-input-4-958facd46620>:1)
Out[4]: True
```

but now that I think about it, users won't use logging in an interactive environment very often and so this shouldn't influence the design for the standard usage.

I will update it to always return the value, thanks

wjwwood commented on rclpy/rclpy/logging.py:
```diff
@@ -0,0 +1,76 @@
+
+
+def get_named_logger(name):
+    return rclpy.impl.logger_rcutils.RcutilsLogger(name)
```
It looks strange to me to have `logger_rcutils` and then `RcutilsLogger` next to it. Why is this not called `rcutils_logger.RcutilsLogger`?

dhood replied:
I imagined there might be other logger implementations and it would be easier if the files were grouped together alphabetically, but we are _so_ far from having that situation! updated in c8807c56c3e6461dd8f8ebfb032dfa497a1aa193

dirk-thomas commented on rclpy/rclpy/impl/logger_rcutils.py:
```diff
@@ -0,0 +1,275 @@
+
+    @classmethod
+    def initialize_context(cls, context, **kwargs):
+        super(cls, cls).initialize_context(context, **kwargs)
```
You can't use `cls` as the fist argument but need to use `Once`. Otherwise if another class subclasses from `Once` this wouldn't work.

Same below.

dhood replied:
argh, thanks. I corrected this in 0c7916c49d178259f1602b5847c96e1d7a509154 but can remove the arguments if we prefer easy copy-paste over explicitness

dhood commented:
> That's a feature that I have pointed out before should be supported by the common logging API (so in rcutils. I believe it wasn't implemented yet because you need some sort of map like structure to correlate settings to a given logger name and we didn't have that in C. I proposed the implementation could use C++ and the std::map structure for the storage and lookup of the settings by logger name, but I think @dirk-thomas wanted to use C and also did not want to implement the data structure. So we left it hanging waiting on a decision about which C library to pull it that would give us that data structure.
You could implement it in python separately, but I think that's wasted effort if we know it needs to work at the rcutils level as well in the future.

@wjwwood Yeah, I was thinking to add logger thresholds in rclpy independent of rcutils just as a temporary measure so that people could leverage it in the interim, but it's not all that useful without an interface for configuring the logger severities. So I'm in favour of leaving the rclpy logger severity thresholds referencing the rcutils global severity threshold for now, and we can update rclpy at the same time we add configuration for loggers in general.

dirk-thomas commented on rclpy/rclpy/impl/logger_rcutils.py:
```diff
@@ -0,0 +1,275 @@
+
+
+# The ordering of this dictionary matches the order in which filters will be processed.
+# Note(dhood): ordering in the constructor is only maintained as of Python 3.6,
```
You could initialize the ordered dict in the constructor in older Python versions, e.g. `OrderedDict([('foo', 1), ('bar', 2), ('baz', 3)])`. The initialization in the following lines is nicer to read though. Therefore I would just remove this comment.

dhood replied:
ah, it was just an issue with how I was using it (I tried to initialised OrderedDict passing a dict, not a list of tuples). I removed the comment in acf11fcf5c98022b9dd0a4bd2703af7ae2fb6183

dirk-thomas commented on rclpy/rclpy/impl/logger_rcutils.py:
```diff
@@ -0,0 +1,275 @@
+                            param_name, detected_filter))
+    for kwarg in kwargs:
+        if kwarg not in all_supported_params:
+            raise RuntimeError(
```
Similar here: unexpected keyword arguments are signaled with a `TypeError`.

dhood commented on rclpy/src/rclpy/_rclpy_logging.c:
```diff
@@ -0,0 +1,129 @@
+{
+  rcutils_logging_initialize();
+  Py_RETURN_NONE;
+}
```
A situation would be that the user wants to register their own output handler or do some other configuration that should take place after initialisation but before logging. Configuration support isn't there at the moment and the initialize function isn't being used in this PR anymore (since https://github.com/ros2/rcutils/pull/52), but I still left it around since it'll be useful eventually. It might be better practice to not add it until it's needed: I can go either way

dhood commented on rclpy/src/rclpy/_rclpy_logging.c:
```diff
@@ -0,0 +1,129 @@
+{
+  int severity = rcutils_logging_get_severity_threshold();
+
+  return PyLong_FromLong(severity);
```
from what I understood the size of a C enum might be an int (maybe that's just in unusual cases though). Either way, long was the smallest option I saw in https://docs.python.org/3/c-api/long.html

dhood commented on rclpy/src/rclpy/_rclpy_logging.c:
```diff
@@ -0,0 +1,129 @@
+ *
+ * \param[in] severity Threshold to set
+ * \return None, or
+ * \return NULL on failure
```
updated in 71ab9a92bb304390bda4a48cf11b20ce9750d58f

mikaelarguedas replied:
Sorry that was unclear in my previous comment, I meant that if we want to document the behavior on failure we should document the error that is raised rather than the return value (so that users can catch it if they want)

dhood replied:
I think I'm missing something. It's not raising an error though, so "None or NULL on failure" seems to be the appropriate documentation for the return value from what I can tell? perhaps what you're suggesting is that it should raise?

mikaelarguedas replied:
PyArg_ParseTuple will raise for you (https://docs.python.org/3.5/c-api/arg.html#c.PyArg_ParseTuple) so the user will never see the return code

dhood replied:
Ah, I see now, thanks. In general we don't document that exception so I agree with your original point that it's not necessary to explain that in the docblock: 226c96e0c7d664aedbbe76a92ba0747fdc393a28

dhood commented on rclpy/src/rclpy/_rclpy_logging.c:
```diff
@@ -0,0 +1,129 @@
+static PyObject *
+rclpy_logging_rcutils_log(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  unsigned PY_LONG_LONG severity;
```
since python could pass in something that big, I thought it made sense (and copying how qos enum is parsed), but I'm open to other suggestions if something else is more appropriate

dhood replied:
After speaking with @mikaelarguedas offline I switched this to int in f6e6b44aff59e549ad981683f533d24be3d59671 and then ensured that the python value being passed through will fit into an int by only ever passing `LoggingSeverity` enums in 9895ca825a5ffa612696731cb61ed95acef97589

dhood commented on rclpy/src/rclpy/_rclpy_logging.c:
```diff
@@ -0,0 +1,129 @@
+    return NULL;
+  }
+
+  RCUTILS_LOGGING_AUTOINIT
```
the macro has been defined without any arguments so AFAIK this is the only usage that compiles

dhood commented on rclpy/src/rclpy/_rclpy_logging.c:
```diff
@@ -0,0 +1,129 @@
+
+  RCUTILS_LOGGING_AUTOINIT
+  rcutils_log_location_t logging_location = {function_name, file_name, line_number};
+  rcutils_log(&logging_location, severity, name, message);
```
`logging_location` isn't an argument but is composed of the three arguments so I moved them to the end, but otherwise the ordering is preserved

dhood commented on rclpy/src/rclpy/_rclpy_logging.c:
```diff
@@ -0,0 +1,129 @@
+   "Get the global severity threshold."},
+  {"rclpy_logging_set_severity_threshold", rclpy_logging_set_severity_threshold, METH_VARARGS,
+   "Set the global severity threshold."},
+  {"rclpy_logging_rcutils_log", (PyCFunction)rclpy_logging_rcutils_log, METH_VARARGS,
```
ah, it _was_ necessary when I was passing keyword arguments (https://stackoverflow.com/questions/10264080/python-c-extension-why-are-methods-that-use-keyword-arguments-cast-to-pycfunct if you're interested), but it's not necessary anymore. removed in 7eaed231f99f855d25d5b17f914f9edccaff59fb

dhood commented on rclpy/src/rclpy/_rclpy_logging.c:
```diff
@@ -0,0 +1,129 @@
+};
+
+PyDoc_STRVAR(rclpy_logging__doc__,
+  "RCLPY module for logging.");
```
I read [here](https://www.python.org/dev/peps/pep-0007/) to: 
> Use the PyDoc_STR() or PyDoc_STRVAR() macro for docstrings to support building Python without docstrings (./configure --without-doc-strings).

so was following that where it was trivial to do so.

mikaelarguedas replied:
Cool thanks for the pointer!

dhood commented on rclpy/CMakeLists.txt:
```diff
@@ -2,6 +2,11 @@ cmake_minimum_required(VERSION 3.5)
+# Default to C11
+if(NOT CMAKE_C_STANDARD)
+  set(CMAKE_C_STANDARD 11)
+endif()
```
I remembered that we went through and added this in a bunch of places and thought that this package got forgotten. Now I see that those PRs were only touching code that had a minimum requirement. This doesn't, so I removed it 13b475c056d4bd9b6719fe2641af7b356b388b3f

dhood commented on rclpy/rclpy/impl/logger_rcutils.py:
```diff
@@ -0,0 +1,275 @@
+
+    params = {
+        'throttle_duration_sec': None,
+        'throttle_time_source_type': 'RCUTILS_STEADY_TIME',
```
the logic is in python but (in theory) it's based on rcutils time underneath. at the moment it's using internal python time but I understood that one day we'd expose rcutils time concepts in rclpy. Even though it's not really using rcutils steady time at the moment, I didn't want to put an honest default because we might not keep support for that. I added an explanation for users changing the option to not expect it to work with other values in 95dbecf9c076c5ac7768c332662b615d02ee5eda

dhood commented on rclpy/rclpy/impl/logger_rcutils.py:
```diff
@@ -0,0 +1,275 @@
+    def log_condition(context):
+        logging_condition = True
+        # TODO(dhood): use rcutils time and the time source type
+        now = time.time()
```
no, I considered wrapping rcutils time concepts an improvement that could be outside of the scope of this PR. I updated the clock type in 3b9a3ce5aa004e2a7c67429082d20ee2c3a9dc0e

dhood commented on rclpy/rclpy/impl/logger_rcutils.py:
```diff
@@ -0,0 +1,275 @@
+            make_log_call &= supported_filters[logging_filter].log_condition(
+                self.contexts[caller_id_str])
+            if not make_log_call:
+                break
```
yeah, that's nicer to read I think. updated in adcc9cde5da7b89e1377c36ffaf0de7d8d6c912e

dhood commented on rclpy/rclpy/logging.py:
```diff
@@ -0,0 +1,76 @@
+# TODO(dhood): document the supported filter options for all logging convenience functions
+# in a way that prevents copy-paste
+def log(message, severity, **kwargs):
+    assert isinstance(severity, LoggingSeverity) or isinstance(severity, int)
```
was basing this off the qos logic but changed to a TypeError in 478f056976042347dcf69daffc22524a4df63861

dhood replied:
update: I removed the TypeError in 9895ca825a5ffa612696731cb61ed95acef97589 in favour of casting to LoggingSeverity, which will complain if the value isn't one of the enum values

dhood commented on rclpy/rclpy/impl/logger_rcutils.py:
```diff
@@ -0,0 +1,275 @@
+
+        # Get/prepare the context corresponding to the caller.
+        caller_id = CallerId()
+        caller_id_str = pickle.dumps(caller_id)
```
otherwise the object is different each time so using it as a key doesn't accomplish the history that's desired. @dirk-thomas suggested overwriting the `__hash__` method: I'll give that a shot

dhood commented on rclpy/rclpy/impl/logger_rcutils.py:
```diff
@@ -0,0 +1,275 @@
+        # Populate the list of known filenames from which logging methods can be called.
+        # This has to be done from within a function to avoid cyclic module imports.
+        import rclpy.logging
+        _internal_callers = [__file__, rclpy.logging.__file__]
```
in 4d97de06660c8d0f2a4e3cd50bedf56cd7f765c2 I added another variable that third-parties can contribute to; I left the check on None for the internal filenames to avoid doing the importing + enforcing each time the function's called

dirk-thomas replied:
Imo having two variables makes it more complex than necessary. If you would rather not check if the items are already in the list you could use a global boolean flag to determine if `_find_caller` should add its own (on the first call) or not (on repeated calls).

dhood replied:
OK, I have change it to use a flag. (most of the change was in cdd4d2ada2a531a4a47226e83b6474efc1319be)

dhood commented on rclpy/rclpy/impl/logger_rcutils.py:
```diff
@@ -0,0 +1,275 @@
+        for param in cls.params:
+            context[param] = kwargs.get(param, cls.params[param])
+            if context[param] is None:
+                raise RuntimeError(
```
thanks for the explanation: faeeabfe9546ed888e1fd85e910f0b6f06560aa1

dhood commented on rclpy/rclpy/impl/logger_rcutils.py:
```diff
@@ -0,0 +1,275 @@
+    Decide if it's appropriate to log given a context, and update the context accordingly.
+    """
+    @staticmethod
+    def log_condition(context):
```
I went with `should_log`: 79fe61d5fd128f3884c49c327eab51ab8ab2ef9a

dirk-thomas commented on rclpy/rclpy/impl/rcutils_logger.py:
```diff
@@ -0,0 +1,286 @@
+    # Check that all required parameters (with no default value) have been specified
+    for detected_filter in detected_filters:
+        for param_name, default_value in supported_filters[detected_filter].params.items():
+            if param_name not in kwargs:
```
To avoid deep nesting I would suggest:

```
if param_name in kwargs:
    continue
```

dhood replied:
nesting avoided in 2c2e9208b7c7dc26382c1a8c00e1e86730cfde8e

dirk-thomas commented on rclpy/rclpy/impl/rcutils_logger.py:
```diff
@@ -0,0 +1,286 @@
+                    raise TypeError(
+                        'required parameter "{0}" not specified '
+                        'but is required for the the logging filter "{1}"'.format(
+                            param_name, detected_filter))
```
This block could also reduce nesting by:

```
if default_value is None:
    raise TypeError(...)
kwargs[param_name] = default_value
```

dirk-thomas commented on rclpy/rclpy/impl/rcutils_logger.py:
```diff
@@ -0,0 +1,286 @@
+            # Call the relevant function from the C extension.
+            _rclpy_logging.rclpy_logging_rcutils_log(
+                severity, name, message,
+                caller_id.function_name, caller_id.file_name, caller_id.line_number)
```
This could be refactored without the boolean variable to:

```
for logging_filter in context['filters']:
    if not supported_filters[logging_filter].should_log(context):
        break
else:
    if severity >= self.get_severity_threshold():
        # Call the relevant function from the C extension.
```

mikaelarguedas replied:
yeah, I think the original reason for not doing so was to have a variable boolean to return. So to use the suggested structure we would need to add return statements in all condition branches, don't we?
```
for logging_filter in context['filters']:
    if not supported_filters[logging_filter].should_log(context):
        return False
else:
    if severity >= self.get_severity_threshold():
        # Call the relevant function from the C extension.
        return True
   else:
        return False
```

dirk-thomas replied:
Oh, I see. You would indeed need a `return True in the `if` block and one `return False` at the very end.

dirk-thomas replied:
This could even be adapted to follow the "return early" philosophy:

```
for logging_filter in context['filters']:
    if not supported_filters[logging_filter].should_log(context):
        return False

if severity < self.get_severity_threshold():
    return False

# Call the relevant function from the C extension.
...
return True
```

dhood replied:
thanks @mikaelarguedas and @dirk-thomas for the pointers! This is much cleaner now: e6a85ca2e4d9793264a1cbe03d595fc25c0e8560

dhood commented:
To fix `_find_caller` on windows, I had to introduce `pathlib.Path().resolve()` wrappers to each of the filenames returned by `inspect.getabsfile()`. This is because `inspect.getabsfile()` appears to return paths  as _lowercase_ on windows (only). Using `pathlib.Path().resolve()` corrects the case somehow (not sure how it would deal with ambiguity?).

Another issue on Windows was that, surprisingly to me, the `__file__` attribute of the module had a mix of forward and backwards slashes.. So I applied the same approach to the internal filenames.

Most of this was done in 6c6917f34ac87b69ae7a8d7c011d61cde9cb752d; if there's a cleaner way that avoids lower casing everything I'm interested.

Example usage on windows is:
```
[INFO] []: message (test_log_arguments() at C:\J\workspace\ci_windows\ws\src\ros2\rclpy\rclpy\test\test_logging.py:128)
```

and on linux:
```
[INFO] []: message (test_log_arguments() at /home/dhood/ros2_ws/src/ros2/rclpy/rclpy/test/test_logging.py:128)
```

dhood commented:
there are still a few outstanding comments that I will respond to tomorrow

dirk-thomas commented on rclpy/rclpy/impl/rcutils_logger.py:
```diff
@@ -0,0 +1,310 @@
+    return frame
+
+
+class CallerId:
```
Maybe the class could use `__slots__` if it is never intended to be extended with additional attributes.

That would also make the implementation of `__hash__` and `__eq__` easier by just iterating the slots and avoiding the repetition of the member names.

dhood replied:
I made use of `__slots__` as you suggested in d34a86cce78a4623cceceda5d66394d760ba121b but then figured what we really want is a named tuple so switched to that in b4a87876cbd5a9d031cb0e280f22cd4369d134d5

dirk-thomas replied:
Even nicer! :1st_place_medal:

dirk-thomas commented on rclpy/rclpy/impl/rcutils_logger.py:
```diff
@@ -0,0 +1,310 @@
+        if not frame:
+            frame = _find_caller(inspect.currentframe())
+        self.function_name = frame.f_code.co_name
+        self.file_name = _normalize_path(inspect.getabsfile(frame))
```
For `file_name` I would only expect the name of the file but not its full path. Maybe rename to `file_path`?

dhood replied:
yeah, fair call: renamed in 5507e4e9b3af2e8cdb15aa5f76c575af5df73f81

dirk-thomas commented on rclpy/rclpy/impl/rcutils_logger.py:
```diff
@@ -0,0 +1,302 @@
+        # This has to be done from within a function to avoid cyclic module imports.
+        import rclpy.logging
+        # Extend the list to preserve any filenames that may have been added by third parties.
+        # Note: the call to `abspath` will also resolve mixed slashes that can result on Windows.
```
Using `realpath` will also resolve symlinks and therefore ensure that they compare equal even if one is a symlink of the other.

dhood replied:
thanks! I changed it in c4ae7a7577f7b4c21d4914562f06f79d69406f44 but only for the `_find_caller` function. the file path that will displayed in the logging call is [still using abspath](https://github.com/ros2/rclpy/blob/c4ae7a7577f7b4c21d4914562f06f79d69406f44/rclpy/rclpy/impl/rcutils_logger.py#L63) because I thought that would be more what users are expecting (but maybe not)

dhood commented:
Indeed using `os.path.abspath` to get the absolute path instead of `inspect.getabsfile` prevents the case change, thanks @dirk-thomas. I switched to that in https://github.com/ros2/rclpy/pull/102/commits/b02824349c133573d8f5182fb24506acc99c97ae
I noted in the code comments that the `abspath` on the internal filenames is needed because otherwise `__file__` can include a mix of slash directions on Windows e.g.  (output as debug print from [this job](http://ci.ros2.org/job/ci_windows/3220/testReport/junit/(root)/projectroot/rclpytests/)):
```
Internal callers: ['C:/J/workspace/ci_windows/ws/src/ros2/rclpy/rclpy\\rclpy\\impl\\rcutils_logger.py', 'C:/J/workspace/ci_windows/ws/src/ros2/rclpy/rclpy\\rclpy\\logging.py']
```


Update on the aforementioned todos:
- Currently all loggers are using the global severity threshold from rcutils: this will stay until we have a way to set thresholds with more granularity in rcutils.
- All of the convenience functions take the same set of filter options but none of them have been documented which inhibits autocompletion: I hard-coded documentation for the main log method and referenced it from the other callers in https://github.com/ros2/rclpy/pull/102/commits/dea180d1c1bd9ac5a63f75a87073d28f18b5b13f. We can avoid the hard-coding of filter options at a later date.
- 'function' and/or 'expression' filter support: post-PR enhancement.
- Allow log calls to take a format string + args (currently only pre-formatted strings are supported): post-PR enhancement for ROS 1 compatibility.
- Use rcutils instead of python time: post-PR enhancement.

With that, I'll put this back in review (I think I've addressed the comments to date, LMK if I missed something)

dhood commented:
:ship: :ship: :tada: 

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=3191)](http://ci.ros2.org/job/ci_linux/3191/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=524)](http://ci.ros2.org/job/ci_linux-aarch64/524/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=2549)](http://ci.ros2.org/job/ci_osx/2549/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=3249)](http://ci.ros2.org/job/ci_windows/3249/)
